subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DefaultFileCollectionResolveContext.java

```diff
@@ -102,50 +103,57 @@ public final ResolvableFileCollectionResolveContext newContext() {
     }
 
     private <T> List<T> doResolve(Converter<? extends T> converter) {
-        List<T> result = new ArrayList<T>();
+        List<T> result = new ArrayList<T>(queue.size());
         while (!queue.isEmpty()) {
-            Object element = queue.remove(0);
-            // TODO - need to sync with BuildDependenciesOnlyFileCollectionResolveContext
-            if (element instanceof DefaultFileCollectionResolveContext) {
-                DefaultFileCollectionResolveContext nestedContext = (DefaultFileCollectionResolveContext) element;
-                converter.convertInto(nestedContext, result, fileResolver);
-            } else if (element instanceof FileCollectionContainer) {
-                FileCollectionContainer fileCollection = (FileCollectionContainer) element;
-                resolveNested(fileCollection, result, converter);
-            } else if (element instanceof FileCollection || element instanceof MinimalFileCollection) {
-                converter.convertInto(element, result, fileResolver);
-            } else if (element instanceof Task) {
-                Task task = (Task) element;
-                queue.add(0, task.getOutputs().getFiles());
-            } else if (element instanceof TaskOutputs) {
-                TaskOutputs outputs = (TaskOutputs) element;
-                queue.add(0, outputs.getFiles());
-            } else if (DeferredUtil.isDeferred(element)) {
-                Object deferredResult = DeferredUtil.unpack(element);
-                if (deferredResult != null) {
-                    queue.add(0, deferredResult);
-                }
-            } else if (element instanceof Path) {
-                queue.add(0, ((Path) element).toFile());
-            } else if (element instanceof Iterable) {
-                Iterable<?> iterable = (Iterable) element;
-                GUtil.addToCollection(queue.subList(0, 0), iterable);
-            } else if (element instanceof Object[]) {
-                Object[] array = (Object[]) element;
-                GUtil.addToCollection(queue.subList(0, 0), Arrays.asList(array));
```

Comment:
this was a bit confusing, i.e. adding multiple elements to the front, in reverse order, to be processed in correct order by the loop. Changed it to process these elements inline, eliminating the need for a double ended queue.

Reply:
This could also bring us closer to making it lazy in the future, since now value generators are separated mostly from the iteration